mobile/src/main/java/ch/epfl/sdp/mobile/ui/speech_recognition/ChessSpeechRecognitionFilter.kt

```diff
@@ -0,0 +1,210 @@
+   * @param confidencesScores list of confidence score for recognized speech
+   * @return chess keyword string after applying all speechRules, or an empty string if all rules
+   * fail
+   */
+  private fun applyRules(
+      speeches: List<List<Word>>,
+      confidencesScores: List<Double>,
+  ): String {
+
+    val candidates = mutableMapOf<String, Double>()
+    for (chessRule in speechRules.rulesSet) {
+      speeches.forEachIndexed { index, speech ->
+        if (chessRule.rule(speech)) {
+          candidates[chessRule.chessPiece] = confidencesScores[index]
+        }
+      }
+    }
+    return bestCandidate(candidates)
+  }
+
+  /**
+   * Returns detected chess keyword candidate with best confidence score
+   * @param candidates mapping from detected in dictionary chess keywords to their scores
+   * @return candidate as a chess keyword
+   */
+  private fun bestCandidate(candidates: Map<Word, Double>): String {
+    if (candidates.isEmpty()) {
+      return ""
+    }
+
+    if (candidates.size == 1) {
+      return candidates.entries.first().key
+    }
+
+    var bestCandidate = candidates.entries.first()
+    for (candidate in candidates) {
+      if (candidate.value > bestCandidate.value) {
+        bestCandidate = candidate
+      }
+    }
+    return bestCandidate.key
+  }
+
+  /**
+   * Detects if any word in the speech is within the numberDictionary and return its equivalent
+   * correct chess piece keyword
+   * @param words a tokenized speech
+   * @param pieceDictionary a mapping from chess piece keywords to their homonyms
+   * @return a chess number keyword or an empty string
+   */
+  private fun detectPieceType(
+      words: List<Word>,
+      pieceDictionary: Map<ChessPiece, List<String>> = chessDictionary.chessPieces
+  ): String {
+    return detect(words, pieceDictionary)
+  }
+
+  /**
+   * Detects if any word in the speech is within the numberDictionary and return its equivalent
+   * correct chess letter keyword
+   * @param words a tokenized speech
+   * @param lettersDictionary a mapping from chess keyword letter to their homonyms
+   * @return a chess number keyword or an empty string
+   */
+  private fun detectLetter(
+      words: List<Word>,
+      lettersDictionary: Map<ChessLetter, List<String>> = chessDictionary.letters
+  ): String {
+    return detect(words, lettersDictionary)
+  }
+
+  /**
+   * Detects if any word in the speech is within the numberDictionary and return its equivalent
+   * correct chess number keyword
+   * @param words a tokenized speech
+   * @param numbersDictionary a mapping from numbers to their homonyms
+   * @return a chess number keyword or an empty string
+   */
+  private fun detectNumber(
+      words: List<Word>,
+      numbersDictionary: Map<ChessNumber, List<String>> = chessDictionary.numbers
+  ): String {
+    return detect(words, numbersDictionary)
+  }
+
+  /**
+   * Makes a decision of choosing a conform by relying on confidence scores
```

Comment:
The documentation is not really clear, can you rewrite this to be more explicit ?

Reply:
Yes !